src/traversals/dfs.jl

```diff
@@ -7,10 +7,28 @@
 Return `true` if graph `g` contains a cycle.
 
 ### Implementation Notes
-Uses DFS.
+The algorithm uses a DFS. Self-loops are counted as cycles.
 """
 function is_cyclic end
-@traitfn is_cyclic(g::::(!IsDirected)) = ne(g) > 0
+@enum Vertex_state unvisited visited
+@traitfn function is_cyclic(g::AG::(!IsDirected)) where {T, AG<:AbstractGraph{T}}
+    visited = falses(nv(g)) 
+    for v in vertices(g)
+        visited[v] && continue
+        visited[v] = true
+        S = [(0,v)] 
+        while !isempty(S)
+            parent, w = pop!(S)
+            for u in neighbors(g, w)
+                u == parent && continue
+                visited[u] && return true
```

Comment:
```suggestion
        S = [(zero(T),v)] 
        while !isempty(S)
            parent, w = pop!(S)
            for u in neighbors(g, w)
                u == parent && continue
                visited[u] && return true
```
otherwise this will lead to type instability. But also, this relies on the assumption that the value `0` is never used as vertex index. While this might indeed be the case, why not write it as

```suggestion
        S = [(v,v)] 
        while !isempty(S)
            parent, w = pop!(S)
            for u in neighbors(g, w)
                u == w && return true # self-loop
                u == parent && continue
                visited[u] && return true
```
?